server/src/internalClusterTest/java/org/elasticsearch/indices/recovery/SnapshotBasedIndexRecoveryIT.java

```diff
@@ -0,0 +1,752 @@
+        String indexName = randomAlphaOfLength(10).toLowerCase(Locale.ROOT);
+        createIndex(indexName,
+            Settings.builder()
+                .put(IndexMetadata.SETTING_NUMBER_OF_SHARDS, 1)
+                .put(IndexMetadata.SETTING_NUMBER_OF_REPLICAS, 0)
+                .put(MergePolicyConfig.INDEX_MERGE_ENABLED, false)
+                .put(IndexService.GLOBAL_CHECKPOINT_SYNC_INTERVAL_SETTING.getKey(), "1s")
+                .build()
+        );
+
+        int numDocs = randomIntBetween(3000, 5000);
+        indexDocs(indexName, 0, numDocs);
+
+        String repoName = "repo";
+        createRepo(repoName, TestRepositoryPlugin.INSTRUMENTED_TYPE);
+        createSnapshot(repoName, "snap", Collections.singletonList(indexName));
+
+        assertAcked(client().admin().indices().prepareDelete(indexName).get());
+
+        List<String> restoredIndexDataNodes = internalCluster().startDataOnlyNodes(2);
+        RestoreSnapshotResponse restoreSnapshotResponse = client().admin().cluster()
+            .prepareRestoreSnapshot(repoName, "snap")
+            .setIndices(indexName)
+            .setIndexSettings(Settings.builder()
+                .put(IndexMetadata.SETTING_NUMBER_OF_REPLICAS, 1)
+                .put("index.routing.allocation.include._name", String.join(",", restoredIndexDataNodes))
+            ).setWaitForCompletion(true)
+            .get();
+
+        RestoreInfo restoreInfo = restoreSnapshotResponse.getRestoreInfo();
+        assertThat(restoreInfo.successfulShards(), is(equalTo(restoreInfo.totalShards())));
+
+        ensureGreen(indexName);
+        assertDocumentsAreEqual(indexName, numDocs);
+
+        RecoveryState recoveryState = getLatestPeerRecoveryStateForShard(indexName, 0);
+        String sourceNode = recoveryState.getSourceNode().getName();
+        String targetNode = recoveryState.getTargetNode().getName();
+
+        assertThat(restoredIndexDataNodes.contains(sourceNode), is(equalTo(true)));
+        assertThat(restoredIndexDataNodes.contains(targetNode), is(equalTo(true)));
+        assertPeerRecoveryWasSuccessful(recoveryState, sourceNode, targetNode);
+
+        // Since we did a restore first, and the index is static the data retrieved by the target node
+        // via repository should be equal to the amount of data that the source node retrieved from the repo
+        InstrumentedRepo sourceRepo = getRepositoryOnNode(repoName, sourceNode);
+        InstrumentedRepo targetRepo = getRepositoryOnNode(repoName, targetNode);
+        assertThat(sourceRepo.totalBytesRead.get(), is(equalTo(targetRepo.totalBytesRead.get())));
+
+        long snapshotSizeForIndex = getSnapshotSizeForIndex(repoName, "snap", indexName);
+
+        assertThat(sourceRepo.totalBytesRead.get(), is(greaterThan(0L)));
+        assertThat(sourceRepo.totalBytesRead.get(), is(lessThanOrEqualTo(snapshotSizeForIndex)));
+    }
+
+    public void testPrimaryHandoverUsesSnapshots() throws Exception {
```

Comment:
I fail to see how this validates more than `testPeerRecoveryUsesSnapshots()`? I probably missed some detail though...

Reply:
🤦 you're right, I've added a new test that adds replicas and stops them randomly instead.